### PR TITLE
add etag middleware

### PR DESCRIPTION
**Pull request by XaoGao**

### CODE REVIEW

**Comment by XaoGao** [comment]
@IvanShamatov please check out etag middleware, 'couse i not quite sure about this

File: lib/course/middleware/assets.rb [hunk 1]
```diff
@@ -36,8 +37,7 @@ def handle_public_request(env)
 
         body = read_file(env)
         status = body == "" ? Statuses::NOT_FOUND : Statuses::SUCSSESS
```
Comment (IvanShamatov) [inline lib/course/middleware/assets.rb]: check spelling of Success here

File: lib/course/middleware/assets.rb [hunk 2]
```diff
@@ -36,8 +37,7 @@ def handle_public_request(env)
...
-        # TODO: add factory to generate contetnt-type
-        [status, { "content-type" => "text/plain" }, [body]]
+        [status, headers(env), [body]]
```
Comment (IvanShamatov) [inline lib/course/middleware/assets.rb]: We should not cache if it failed

File: lib/course/middleware/assets.rb [hunk 3]
```diff
@@ -57,6 +57,15 @@ def full_path(env)
       def danger_path?(env)
         env["REQUEST_PATH"].sub(PUBLIC_URL, "").include?("../")
       end
+
+      def headers(env)
+        header = {}
+        header["content-type"] = env["CONTENT_TYPE"] || "text/plain"
```
Comment (IvanShamatov) [inline lib/course/middleware/assets.rb]: I'm not sure, why you want to specify a content type here, but for assets, which are usually files css/js/png/pdf/jpg/csv/blah_blah, text/plain would not be the best default value.

File: lib/course/middleware/assets.rb [hunk 4]
```diff
@@ -57,6 +57,15 @@ def full_path(env)
...
+        header["cache-control"] = "public, max-age=#{MAX_CACHE_AGE}"
+        header["expires"] = (Time.now + MAX_CACHE_AGE).utc.rfc2822
+
+        header
```
Comment (IvanShamatov) [inline lib/course/middleware/assets.rb]: can be rewritten nicer, like that:
```ruby
def headers(env)
  {
    "cache-control" => "public, max-age=#{MAX_CACHE_AGE}",
    "expires" => (Time.now + MAX_CACHE_AGE).utc.rfc2822
  }
end
```

File: lib/course/middleware/etag.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
+module Course
+  module Middleware
+    class Etag
+      attr_reader :app
+
+      def initialize(app)
+        @app = app
+      end
+
+      def call(env)
```
Comment (IvanShamatov) [inline lib/course/middleware/etag.rb:12]: no, you got it wrong
please read carefully how etag works.
Yes we need to calculated digest, this is correct.

PS. This middleware expect `digest` lib to be required, we should put it right here, not elsewhere.

**Comment by XaoGao** [comment]
@IvanShamatov fix etag middleware, please take a look

File: lib/course/middleware/etag.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+          return [status, headers, body]
+        end
+
+        [status, headers, body]
+      end
+
+      private
+
+      def success_status?(status)
+        status == Statuses::SUCSSESS
```
Comment (IvanShamatov) [inline lib/course/middleware/etag.rb]: SPELLING

File: lib/course/middleware/etag.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      end
+
+      def cache?(headers)
+        headers.include? "cache-control"
+      end
+
+      # body is an array
+      def digest_body(body)
+        body.filter { |item| !item.empty? }
```
Comment (IvanShamatov) [inline lib/course/middleware/etag.rb]: why not just hash the whole body?

**Review by IvanShamatov — APPROVED** [review]
nice, just fix the spelling SUCCESS, and body digesting

File: lib/course/middleware/etag.rb (hunk 7 above)
Comment (XaoGao) [reply lib/course/middleware/etag.rb]: i did not understand the question, param for Digest::MD5.hexdigest  have to string or question about filter? i though no point for cache nil or empty string

**Comment by XaoGao** [comment]
@IvanShamatov fixed spelling, but have question https://github.com/XaoGao/rack-middleware/pull/5#discussion_r1133243813

File: lib/course/middleware/etag.rb (hunk 7 above)
Comment (IvanShamatov) [reply lib/course/middleware/etag.rb]: Though there is no harm in caching nil, and md5(body) would be easier to write, read and understand. Though you can filter much prittier: body.compact

**Pull request merged**